### ISSUE TITLE
Add support for specifying the order of nested statements

### ISSUE THREAD

**Issue by alan-agius4** [stylelint/stylelint#973]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

Add rule caring about declaration ordering of rule sets example @extend declarations, @include declarations without inner @content, properties, @include declarations with inner @content, then nested rule sets.

> Which rule, if any, is this issue related to?

`declaration-block-order`

> What CSS is needed to reproduce this issue?
```css
.error {
    p {
        ...
    }
    
    color: #fff;
    @include my-mixin();
    @extend %placeholder;
}
```

> the above should return an warning when setting the below settings, while the one below should be valid

```css
.error {
    @extend %placeholder;
    @include my-mixin();
    color: #fff;

    p {
        ...
    }
}
```
> What stylelint configuration is needed to reproduce this issue?

```json
{
  "rules": {
    "declaration-block-order": ['extend','include', 'properties', 'pseudo','nested']
  }
}
```

**Comment by davidtheclark** [comment]
A more accurate title might be `nested-statement-order` (since at-rules and rules are both "statements").

**Comment by alan-agius4** [comment]
Yeah, seems good.

**Comment by jeddy3** [comment]
I'm going to mark this as `status: discussion` as I think this is a going to be complex rule (perhaps surpassing the complexity of `declaration-block-properties-order`), and so we'll need to thrash out exactly how this rule is going to work.

Off hand I can think of one approach that could get us partly there and is based on @evilebottnawi [suggestion a while ago](https://github.com/stylelint/stylelint/issues/550#issuecomment-162950322) for dealing with an indefinite list of at-rule types:

```json
{
"block-nested-node-order":
    [
       {
            "type": "at-rule",
            "name": "extend"
        },
        {
            "type": "at-rule",
            "name": "include"
        },
        {
            "type": "decl"
        },
        {
            "type": "rule"
        },
    ]
}
```

But this doesn't account for things like " @include declarations without inner @content [vs] @include declarations with inner @content" or " "pseudo"'s.

Everyone feel free to throw in whatever ideas they might have, however radical, as I don't think there's much like this rule in stylelint. 

P.S. We could look to  [`postcss-sorting`](https://github.com/hudochenkov/postcss-sorting#postcss-sorting-) for inspiration, but this also seems to have the same limitations. Also, it's worth noting that `postcss-sorting` merges sorting declarations (by property name) and statements. Perhaps we should considering merging this rule with `declaration-block-properties-order`?

**Comment by davidtheclark** [comment]
> Perhaps we should considering merging this rule with declaration-block-properties-order?

That might be a good idea. Especially if it turns out that this rule would have to duplicate some of the logic in that one.

**Comment by alan-agius4** [comment]
Seems like a good idea to me too and from the example of the rule shown up here. Seems neat

**Comment by jeddy3** [comment]
@alan-agius4 Is this a feature you'd like to champion i.e. help build and maintain? It's a significant undertaking and we'd appreciate any help driving this forward.

**Comment by alan-agius4** [comment]
I'll try to give it a shot, though I am a bit busy at the moment :), I'll do my best

**Comment by programbo** [comment]
The `style-lint` rule for ordering at-rules is basic but works very well. Put at-rules at the top _unless_ they have a content block. e.g.

```scss
.thing {
  @include grid-row;
  @include font(leto, bold);
  display: block;
  @include breakpoint(large) {
    display: none;
  }
}
```

**Comment by davidtheclark** [comment]
Based on recent discussions (e.g. https://github.com/stylelint/stylelint/issues/1308) I'm thinking this should be implemented in a plugin, not in core. It mostly involves non-standard syntax, and it's wide open to bugs and preferential quirks. @stylelint/core what do you think?

**Comment by evilebottnawi** [comment]
@davidtheclark i think we should have basic plugin in core for order this, without considering non standard syntax.

**Comment by davidtheclark** [comment]
@evilebottnawi Would this then be the same as #962?

**Comment by evilebottnawi** [comment]
@davidtheclark no, this rule disable use all except pseudo, here we speak about order `at-rules`, `pseudo-classes`, `pseudo-element` and etc.

**Comment by davidtheclark** [comment]
> this rule disable use all except pseudo

Not sure what this means ... could you reword?

Nested at-rules are not standard, right? Except for `@nest` which should be treated like other nested rules, I'd think.

**Comment by evilebottnawi** [comment]
@davidtheclark we can write more generic rule and this rule work fine with nested.
```css
.a { 
}

.a:hover,
.a:focus {
}

@media screen and (min-width: 100px) {
.a { }
}
```
and etc. It is really help to read another code and control other developers

**Comment by jeddy3** [comment]
See: https://github.com/stylelint/stylelint/issues/1308
